packages/gatsby/src/query/query-watcher.ts

```diff
@@ -134,11 +134,14 @@ const watch = async (rootDir: string): Promise<void> => {
     return slash(path.join(module.path, pathRegex))
   })
 
+  // If we watch only JS and TS files with a glob pattern, deleting a
+  // directory with a file in it is not caught here and pollutes async-requires.
+  // Subsequent changes then causes a JavaScript bundle failure.
+  // TODO Explore better solutions and the memory implications of this change
```

Comment:
The memory implications of this is my major concern. I think the approach you have looks good overall, I'm just afraid that larger projects are going to see largely adverse effects of this change. If only glob had support of `"{some-glob}/.."`...

Another (probably really bad) option would be to build up a list of directories that contain our desired file extensions. I imagine that would have even poorer performance that what's laid out here, though.

Reply:
Agreed, in fact this by itself isn't enough to fix the issue anyway. Was able to narrow it down to the store still having a reference to the deleted file when we rebuild the JS bundle. Given the complexity in the develop/wait state machine will need to spend some time there.